File: packages/web/src/utils/documentAnalysisSummary.ts
export interface DocumentExtracted {
  documentCategory?: string;
  serviceType?: string;
  totalCost?: number;
  serviceDate?: string;
  mileage?: number;
}

export function getSourceSnippet(
  sourceText: string | undefined,
  maxLength = 240
): string {
  if (!sourceText) return '';
  const compact = sourceText.replace(/\s+/g, ' ').trim();
  if (compact.length <= maxLength) return compact;
  return `${compact.slice(0, maxLength)}...`;
}

export function buildDocumentSummary(
  extracted: DocumentExtracted | undefined,
  sourceText?: string
): string {
  if (!extracted) {
    const snippet = getSourceSnippet(sourceText, 120);
    return snippet || 'No analysis summary available yet';
  }

  const category = extracted.documentCategory
    ? extracted.documentCategory.replace(/_/g, ' ')
    : 'document';

  const details: string[] = [];
  if (extracted.serviceType) {
    details.push(extracted.serviceType);
  }
  if (typeof extracted.totalCost === 'number') {
    details.push(`$${extracted.totalCost.toFixed(2)}`);
  }
  if (extracted.serviceDate) {
    details.push(extracted.serviceDate);
  }
  if (typeof extracted.mileage === 'number') {
    details.push(`${extracted.mileage.toLocaleString()} mi`);
  }

  if (details.length > 0) {
    return `${titleCase(category)}: ${details.join(' • ')}`;
  }

  const snippet = getSourceSnippet(sourceText, 120);
  if (snippet) {
    return `${titleCase(category)}: ${snippet}`;
  }

  return `${titleCase(category)}: No key fields extracted yet`;
}

function titleCase(value: string): string {
  return value
    .split(' ')
    .filter(Boolean)
    .map(token => token.charAt(0).toUpperCase() + token.slice(1).toLowerCase())
    .join(' ');
}

File: packages/mobile/lib/utils/document_analysis_summary.dart
// Mobile port of packages/shared/src/documentAnalysisSummary.ts.

import 'number_format.dart';

String getSourceSnippet(String? sourceText, {int maxLength = 240}) {
  if (sourceText == null || sourceText.isEmpty) return '';
  final compact = sourceText.replaceAll(RegExp(r'\s+'), ' ').trim();
  if (compact.length <= maxLength) return compact;
  return '${compact.substring(0, maxLength)}...';
}

String _titleCase(String value) {
  return value
      .split(' ')
      .where((token) => token.isNotEmpty)
      .map(
        (token) =>
            token.substring(0, 1).toUpperCase() +
            token.substring(1).toLowerCase(),
      )
      .join(' ');
}

String buildDocumentSummary(
  Map<String, dynamic>? extracted,
  String? sourceText, {
  String? locale,
}) {
  if (extracted == null) {
    final snippet = getSourceSnippet(sourceText, maxLength: 120);
    return snippet.isNotEmpty ? snippet : 'No analysis summary available yet';
  }

  final categoryRaw = extracted['documentCategory'];
  final category = (categoryRaw is String && categoryRaw.isNotEmpty)
      ? categoryRaw.replaceAll('_', ' ')
      : 'document';

  final details = <String>[];
  final serviceType = extracted['serviceType'];
  if (serviceType is String && serviceType.isNotEmpty) {
    details.add(serviceType);
  }
  final totalCost = extracted['totalCost'];
  if (totalCost is num) {
    details.add(formatCurrencyAmount(totalCost, locale: locale));
  }
  final serviceDate = extracted['serviceDate'];
  if (serviceDate is String && serviceDate.isNotEmpty) {
    details.add(serviceDate);
  }
  final mileage = extracted['mileage'];
  if (mileage is num) {
    details.add('${formatWithCommas(mileage, locale: locale)} mi');
  }

  if (details.isNotEmpty) {
    return '${_titleCase(category)}: ${details.join(' • ')}';
  }

  final snippet = getSourceSnippet(sourceText, maxLength: 120);
  if (snippet.isNotEmpty) {
    return '${_titleCase(category)}: $snippet';
  }

  return '${_titleCase(category)}: No key fields extracted yet';
}
